codbex-hestia/api/ProductService.ts

```diff
@@ -24,14 +28,44 @@ class ProductService {
         const activeProducts = allProducts.filter(product => product.Enabled === true).length;
         const inactiveProducts = allProducts.filter(product => product.Enabled === false).length;
 
-
         const activeCategories: number = this.categoryDao.count();
 
+        const allSalesInvoiceItems = this.salesInvoiceItemDao.findAll();
```

Comment:
Use the filter functionality to fetch only the relevant records, you can see examples in `OrderService` and `InvoiceService`

Reply:
The performance will be even better if you switch to SQL query with grouping, filtering, sorting, joining, etc.
Consider using SQL instead of DAOs.

**Pros**: performance is better which is important in the current case where you have dashboard with a lot of different widgets(queries). OOM prevention. I can achieve complex analyses easier.
**Cons**: SQL query may be specific for a particular Database type.

Reply:
For some cases DAOs are valid and maybe we need some guidelines, lets say:

- no aggregations and joins -> DAO
- more complex logic like join, aggregations, etc -> query
- for some cases maybe there will be compromises between the two

For queries we should use: https://www.dirigible.io/api/database/query/ as its easier to use.

Thoughts ?
@MinaDoncheva @iliyan-velichkov

Reply:
> For some cases DAOs are valid and maybe we need some guidelines, lets say:
> 
> * no aggregations and joins -> DAO
> * more complex logic like join, aggregations, etc -> query
> * for some cases maybe there will be compromises between the two
> 
> For queries we should use: https://www.dirigible.io/api/database/query/ as its easier to use.
> 
> Thoughts ? @MinaDoncheva @iliyan-velichkov

Sounds good to me. In Electra, I used the query API only - https://github.com/codbex/codbex-electra/blob/main/codbex-electra/ui/api/DashboardService.ts
Another way to decide between the two is - if there is additional processing after the DAO `findAll`, most probably it is better to use the query API. 
Most probably, when we migrate to the Dirigible reporting functionality, we will not have to think about this :)